### PR TITLE
feat(cli): New `routes` commant printing all routes

### LINKED ISSUE #103: Add a routes cli command
that, like Rails, displays all routes that Denali responds to.

### PR DESCRIPTION
**Pull request by knownasilya**
`denali routes` will traverse all addons and your app and print the
endpoint, method and the action.

Fixes #103 

cc @acorncom 

screen shot 2016-09-28 at 6 47 58 pm

### CODE REVIEW

File: lib/runtime/addon.js [hunk 1]
```diff
@@ -157,8 +157,7 @@ export default class Addon {
     forIn(this.pkg.dependencies, (version, pkgName) => {
       let pkgMainPath = resolve.sync(pkgName, { basedir: this.dir });
       let pkgJSONPath = path.resolve(findup('package.json', { cwd: pkgMainPath }));
-      // TODO this shouldn't need to be aware of dist
-      let pkgDir = path.join(path.dirname(pkgJSONPath), 'dist');
+      let pkgDir = path.dirname(pkgJSONPath);
```
Comment (knownasilya) [inline lib/runtime/addon.js]: Bug fix

File: lib/utils/topsort.js [hunk 2]
```diff
@@ -7,7 +7,7 @@ export default function topsort(items, options = {}) {
     graph.add(item.name, value, item.before, item.after);
   });
   let sorted = [];
-  graph.topsort(({ value }) => {
+  graph.topsort((key, value) => {
```
Comment (knownasilya) [inline lib/utils/topsort.js]: Bug fix

File: lib/cli/commands/routes.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      let application = new UserApplication({
+        environment: process.env.DENALI_ENV || process.env.NODE_ENV || 'development',
+        dir: path.join(process.cwd(), 'dist')
+      });
+
+      return application.runInitializers().then(() => {
+        let routes = application.router.routes;
+        let methods = Object.keys(routes);
+        let table = new CliTable({
+          head: [ 'URI', 'ACTION' ]
```
Comment (davewasmer) [inline lib/cli/commands/routes.js]: URI => URL?

File: lib/runtime/addon.js (hunk 1 above)
Comment (davewasmer) [inline lib/runtime/addon.js]: See my comments on #97 - not sure this is a bug.

Comment (knownasilya) [inline lib/runtime/addon.js]: Otherwise the url has `dist/dist/...` and doesn't find the file.

File: lib/cli/commands/routes.js (hunk 3 above)
Comment (knownasilya) [inline lib/cli/commands/routes.js]: Yup, was just copying the rails setup, but I like `URL` better.

**Comment by knownasilya** [comment]
> your screenshot above seems to show a route on the first row of the table that is missing a path - is that intentional?

The routes has a duplicate route, one with `spec: ''` but linking to the same action. A bug in the router?

File: lib/runtime/addon.js (hunk 1 above)
Comment (davewasmer) [inline lib/runtime/addon.js]: Saw your comments on #97, makes sense now. Thanks for clarifying!

Comment (acorncom) [inline lib/runtime/addon.js]: 👍

**Comment by knownasilya** [comment]
So yes there is a duplicate route created, one with a slash and the other without.

screen shot 2016-09-29 at 1 55 36 pm

**Comment by knownasilya** [comment]
This should be ready to merge @davewasmer

**Comment by davewasmer** [comment]
Awesomesauce @knownasilya - thanks!

**Pull request merged**